src/etl/impl/GrpcSource.cpp

```diff
@@ -0,0 +1,152 @@
+        ss << endpoint;
+        grpc::ChannelArguments chArgs;
+        chArgs.SetMaxReceiveMessageSize(-1);
+        stub_ = org::xrpl::rpc::v1::XRPLedgerAPIService::NewStub(
+            grpc::CreateCustomChannel(ss.str(), grpc::InsecureChannelCredentials(), chArgs)
+        );
+        LOG(log_.debug()) << "Made stub for remote.";
+    } catch (std::exception const& e) {
+        LOG(log_.warn()) << "Exception while creating stub: " << e.what() << ".";
+    }
+}
+
+std::pair<grpc::Status, org::xrpl::rpc::v1::GetLedgerResponse>
+GrpcSource::fetchLedger(uint32_t sequence, bool getObjects, bool getObjectNeighbors)
+{
+    org::xrpl::rpc::v1::GetLedgerResponse response;
+    if (!stub_)
+        return {{grpc::StatusCode::INTERNAL, "No Stub"}, response};
+
+    // Ledger header with txns and metadata
+    org::xrpl::rpc::v1::GetLedgerRequest request;
+    grpc::ClientContext context;
+
+    request.mutable_ledger()->set_sequence(sequence);
+    request.set_transactions(true);
+    request.set_expand(true);
+    request.set_get_objects(getObjects);
+    request.set_get_object_neighbors(getObjectNeighbors);
+    request.set_user("ETL");
+
+    grpc::Status const status = stub_->GetLedger(&context, request, &response);
+
+    if (status.ok() && !response.is_unlimited()) {
+        log_.warn() << "is_unlimited is false. Make sure secure_gateway is set correctly on the ETL source. Status = "
+                    << status.error_message();
+    }
+
+    return {status, std::move(response)};
+}
+
+std::pair<std::vector<std::string>, bool>
+GrpcSource::loadInitialLedger(uint32_t const sequence, uint32_t const numMarkers, bool const cacheOnly)
+{
+    if (!stub_)
+        return {{}, false};
+
+    std::vector<etl::impl::AsyncCallData> calls = impl::makeAsyncCallData(sequence, numMarkers);
+
+    LOG(log_.debug()) << "Starting data download for ledger " << sequence << ".";
+
+    grpc::CompletionQueue cq;
+    for (auto& c : calls)
+        c.call(stub_, cq);
+
+    void* tag = nullptr;
+    bool ok = false;
+    size_t numFinished = 0;
+    bool abort = false;
+    size_t const incr = 500000;
+    size_t progress = incr;
+    std::vector<std::string> edgeKeys;
+
+    while (numFinished < calls.size() && cq.Next(&tag, &ok)) {
+        ASSERT(tag != nullptr, "Tag can't be null.");
+        auto ptr = static_cast<etl::impl::AsyncCallData*>(tag);
+
+        if (!ok) {
+            LOG(log_.error()) << "loadInitialLedger - ok is false";
+            return {{}, false};  // handle cancelled
+        }
+
+        LOG(log_.trace()) << "Marker prefix = " << ptr->getMarkerPrefix();
+
+        auto result = ptr->process(stub_, cq, *backend_, abort, cacheOnly);
+        if (result != etl::impl::AsyncCallData::CallStatus::MORE) {
+            ++numFinished;
+            LOG(log_.debug()) << "Finished a marker. "
+                              << "Current number of finished = " << numFinished;
+
+            std::string const lastKey = ptr->getLastKey();
```

Comment:
auto can be used here , I think it has been like this for long. Maybe we can put this line inside the if.